### DISCUSSION TITLE
Option to assign node colours in a group with hash of the group's name

### DISCUSSION THREAD

**Discussion by GrimPixel** [IndiePython/nodezator#59] [Ideas]
Nodes from the same group, like “formula” or “geometric drawings” in “mynodepack”, can be assigned a same colour. 
There is a package for this: <s>[color-hash-python](https://github.com/dimostenis/color-hash)</s>, corrected:  [color-hash-python](https://github.com/dimostenis/color-hash-python)

**Comment by KennedyRichard** [comment]
The link isn't working, but I managed to find the repo you mentioned by looking for the name on google.

That is not a bad idea per se, but the problem is that we can't allow any color to be used, because a requirement of the color is that it must contrast nicely with the color of the text in the title of the node. Even if that package assigns a beautiful color that is too different, if the brightness is close enough, the text might become hard to read.

Because of that, the colors are predefined and the user must pick one of the existing colors offered when changing the color in the menubar **Graph > change category colors**. Even so, I do want users to have freedom to choose the colors in the entire app:

- the colors of the nodes;
- the colors of the background;
- the colors on the UI;
- the colors on the sockets;
- the colors in syntax-highlighted Python code;
- etc.

But for that I must first implement a theme system or something similar. This way users will be able to create and share their favorite themes with the colors they chose. People will be able to pick colors that go well together not only aesthetically, but in a way that doesn't harm readability as well.

In fact, since the beginning, I've been developing Nodezator in a way that is easy to redefine the colors on the app, so a ready theme system is already being formed, I just need to add more stuff.

For instance, you can see the current colors in this folder https://github.com/IndiePython/nodezator/tree/main/nodezator/data/app_themes. The current colors in the app is in the file emeralds_on_coal.pyl, which is the name of the default theme that I created. This file is just a Python file holding the different colors.

You can also see available themes for syntax-highlighted Python code here in this folder: https://github.com/IndiePython/nodezator/tree/main/nodezator/data/syntax_themes/python. The default python syntax theme is the sprinkled_dark.pyl.

I just didn't have the time to write UI to help users choose and define their own themes. That's because there's other features more important and urgent to be implemented, like group nodes (subgraphs). I also didn't do that yet because there's so much to implement/change in Nodezator, that the new elements will be added and existing ones might be removed/changes, so the themes would become obsolete very quickly, since there's always new stuff being implemented. Once Nodezator is more stable it will be more easy to tend to this stuff.

But again, thank you for the idea, feel free to share as many as you have, as soon as they pop up. They are much appreciated and one of them might make into the app, we only know after sharing them.

**Reply by GrimPixel** [reply]
Thank you for the details. I think [distinctipy](https://github.com/alan-turing-institute/distinctipy) can guarantee the high contrast. The motivation of having automatic colour generation as an option is to save the efforts of setting different colours manually.

**Reply by KennedyRichard** [reply]
The idea is super fine, the only real problem is having to use a third-party library to do the job. I'd like to keep Nodezator as external dependency-free as possible, as simple as possible, even more for something like this which  is a "nice to have" rather than a "must have".

But I'll leave this discussion open, at least for the time being, since it is something to keep in mind and a nice idea, overall.
